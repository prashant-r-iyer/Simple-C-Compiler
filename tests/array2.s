	.text

.globl main
main:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9

	# push 8
	movq $8, %rbx

	# push 20
	movq $20, %r10

	# *
	imulq %r10, %rbx
	# func=malloc nargs=1

	# Move values from reg stack to reg args
	movq %rbx, %rdi
	call malloc
	movq %rax, %rbx
	movq %rbx, -8(%rbp)

	# push 0
	movq $0, %rbx
	movq %rbx, -16(%rbp)

for_condition_1:
	movq -16(%rbp), %rbx

	# push 20
	movq $20, %r10
	movq $0, %rax
	cmpq %r10, %rbx
	setl %al
	movq %rax, %rbx
	cmpq $0, %rbx
	je for_end_1
	jmp for_statement_1

for_increment_1:
	movq -16(%rbp), %rbx

	# push 1
	movq $1, %r10

	# +
	addq %r10, %rbx
	movq %rbx, -16(%rbp)
	jmp for_condition_1

for_statement_1:
	movq -16(%rbp), %rbx

	# push 3
	movq $3, %r10
	movq -16(%rbp), %r13

	# *
	imulq %r13, %r10
	imulq $8, %rbx
	addq -8(%rbp), %rbx
	movq %r10, (%rbx)
	jmp for_increment_1

for_end_1:
	movq $string0, %rbx
	# func=printf nargs=1

	# Move values from reg stack to reg args
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx

	# push 0
	movq $0, %rbx
	movq %rbx, -16(%rbp)

for_condition_2:
	movq -16(%rbp), %rbx

	# push 20
	movq $20, %r10
	movq $0, %rax
	cmpq %r10, %rbx
	setl %al
	movq %rax, %rbx
	cmpq $0, %rbx
	je for_end_2
	jmp for_statement_2

for_increment_2:
	movq -16(%rbp), %rbx

	# push 1
	movq $1, %r10

	# +
	addq %r10, %rbx
	movq %rbx, -16(%rbp)
	jmp for_condition_2

for_statement_2:
	movq $string1, %rbx
	movq -16(%rbp), %r10
	movq -16(%rbp), %r13
	imulq $8, %r13
	addq -8(%rbp), %r13
	movq (%r13), %r13
	# func=printf nargs=3

	# Move values from reg stack to reg args
	movq %r13, %rdx
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	jmp for_increment_2

for_end_2:

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

string0:
	.string "Ok so far\n"

string1:
	.string "%d: %d\n"

