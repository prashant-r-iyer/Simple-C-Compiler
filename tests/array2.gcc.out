array2.c: In function ‘main’:
array2.c:8:13: warning: implicit declaration of function ‘malloc’ [-Wimplicit-function-declaration]
    8 |         a = malloc(8*20);
      |             ^~~~~~
array2.c:1:1: note: include ‘<stdlib.h>’ or provide a declaration of ‘malloc’
  +++ |+#include <stdlib.h>
    1 | 
array2.c:8:13: warning: incompatible implicit declaration of built-in function ‘malloc’ [-Wbuiltin-declaration-mismatch]
    8 |         a = malloc(8*20);
      |             ^~~~~~
array2.c:8:13: note: include ‘<stdlib.h>’ or provide a declaration of ‘malloc’
array2.c:14:9: warning: implicit declaration of function ‘printf’ [-Wimplicit-function-declaration]
   14 |         printf("Ok so far\n");
      |         ^~~~~~
array2.c:1:1: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
  +++ |+#include <stdio.h>
    1 | 
array2.c:14:9: warning: incompatible implicit declaration of built-in function ‘printf’ [-Wbuiltin-declaration-mismatch]
   14 |         printf("Ok so far\n");
      |         ^~~~~~
array2.c:14:9: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
array2.c:17:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
   17 |                 printf("%d: %d\n", i, a[i]);
      |                         ~^         ~
      |                          |         |
      |                          int       long int
      |                         %ld
array2.c:17:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘long int’ [-Wformat=]
   17 |                 printf("%d: %d\n", i, a[i]);
      |                             ~^        ~~~~
      |                              |         |
      |                              int       long int
      |                             %ld
